### PR TITLE
Upgrade the DontRepeatTypeInStaticProperties rule.

### PR DESCRIPTION
**Pull request by shawnhyam**
- Using the `visit` method to examine all the variable declarations should allow specific decls to be ignored.
- Update the PipelineGenerator to call `visitPost` methods
- Modify the PipelineGenerator to produce output matching the current generated files.

### CODE REVIEW

**Comment by shawnhyam** [comment]
Should be a partial fix for #771, since it upgrades one rule only. Interested in feedback on if there is a better way to do this before upgrading any other rules.

File: Sources/SwiftFormat/Core/Pipelines+Generated.swift [hunk 1]
```diff
@@ -76,17 +76,16 @@ class LintPipeline: SyntaxVisitor {
...
     visitIfEnabled(BeginDocumentationCommentWithOneLineSummary.visit, for: node)
-    visitIfEnabled(DontRepeatTypeInStaticProperties.visit, for: node)
     visitIfEnabled(NoLeadingUnderscores.visit, for: node)
     visitIfEnabled(TypeNamesShouldBeCapitalized.visit, for: node)
     visitIfEnabled(UseTripleSlashForDocumentationComments.visit, for: node)
     return .visitChildren
   }
   override func visitPost(_ node: ClassDeclSyntax) {
     onVisitPost(rule: AllPublicDeclarationsHaveDocumentation.self, for: node)
+    visitPost(AlwaysUseLowerCamelCase.visitPost, for: node)
```
Comment (ahoppen) [inline Sources/SwiftFormat/Core/Pipelines+Generated.swift:86]: Is this an unrelated change? And should this be `onVisitPost`. I don’t actually know but `visitPost` sticks out to me here.

File: Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift [hunk 2]
```diff
@@ -110,3 +67,38 @@ extension Finding.Message {
     "remove the suffix '\(type)' from the name of the variable '\(name)'"
   }
 }
+
+extension Syntax {
+  /// Returns the name of the immediately enclosing type of this node if there is one,
+  /// otherwise nil.
+  fileprivate func typeName() -> String? {
+    switch Syntax(self).as(SyntaxEnum.self) {
+    case .classDecl(let node):
```
Comment (ahoppen) [inline Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift:76]: You should also check for `.actorDecl`.

File: Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift [hunk 3]
```diff
@@ -110,3 +67,38 @@ extension Finding.Message {
     "remove the suffix '\(type)' from the name of the variable '\(name)'"
   }
 }
+
+extension Syntax {
+  /// Returns the name of the immediately enclosing type of this node if there is one,
+  /// otherwise nil.
+  fileprivate func typeName() -> String? {
```
Comment (ahoppen) [inline Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift:74]: This could be a computed property. Seems more swifty to me.

```suggestion
  fileprivate var typeName: String? {
```

File: Sources/generate-swift-format/PipelineGenerator.swift [hunk 4]
```diff
@@ -93,12 +94,20 @@ final class PipelineGenerator: FileGenerator {
         handle.write(
           """
             override func visitPost(_ node: \(nodeType)) {
+          
           """
         )
-        for ruleName in lintRules.sorted() {
+       for ruleName in lintRules {
```
Comment (ahoppen) [inline Sources/generate-swift-format/PipelineGenerator.swift:100]: Unintentionally removed an indentation whitespace here?

File: Sources/SwiftFormat/Core/Pipelines+Generated.swift (hunk 1 above)
Comment (shawnhyam) [reply Sources/SwiftFormat/Core/Pipelines+Generated.swift:86]: My original implementation used `visitPost`, but there was a problem because the generated pipeline doesn't pass on `visitPost` calls to rules that override those methods. `onVisitPost` only does some cleanup of the pipeline. So the code would work as a standalone visitor, but not as part of the pipeline. I ended up changing the implementation to not use `visitPost`, but I did keep the fix to the code generator in.

Comment (ahoppen) [reply Sources/SwiftFormat/Core/Pipelines+Generated.swift:86]: OK, I’m not too familiar with how the entire pipeline works but I trust you on this 😉

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@shawnhyam Instead of merging `main` into your branch, could you rebase your branch? I find it creates a nicer git history.

**Comment by shawnhyam** [comment]
@allevato I was wondering if you had any thoughts on this, especially the changes to the pipeline generator (so that `visitPost` gets called).

File: Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift [hunk 5]
```diff
@@ -110,3 +67,40 @@ extension Finding.Message {
     "remove the suffix '\(type)' from the name of the variable '\(name)'"
   }
 }
+
+extension Syntax {
+  /// Returns the name of the immediately enclosing type of this node if there is one,
+  /// otherwise nil.
+  fileprivate var typeName: String? {
```
Comment (allevato) [inline Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift:74]: The name of this property feels a little misleading; when I saw you accessing the `typeName` of a `VariableDeclSyntax` node above, it read to me like it was accessing the type of the *variable*. Can we call it `containingDeclName` so that the nature of the traversal is clear at the call site?

File: Sources/SwiftFormat/Rules/NeverForceUnwrap.swift [hunk 6]
```diff
@@ -16,6 +16,7 @@ import SwiftSyntax
 ///
 /// This rule does not apply to test code, defined as code which:
 ///   * Contains the line `import XCTest`
+///   * The function is marked with `@Test` attribute
```
Comment (allevato) [inline Sources/SwiftFormat/Rules/NeverForceUnwrap.swift:19]: Is this noise from another PR that was merged in?

File: Sources/SwiftFormat/Core/Pipelines+Generated.swift (hunk 1 above)
Comment (allevato) [reply Sources/SwiftFormat/Core/Pipelines+Generated.swift:86]: Can we drop this and the code generator changes then, if they're not being used?

File: Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift [hunk 7]
```diff
@@ -22,69 +22,26 @@ import SwiftSyntax
...
+      guard let identifierPattern = binding.pattern.as(IdentifierPatternSyntax.self) else {
+        continue
+      }
 
-        let varName = identifierPattern.identifier.text
-        if varName.contains(bareTypeName) {
-          diagnose(.removeTypeFromName(name: varName, type: bareTypeName), on: identifierPattern)
-        }
+      let varName = identifierPattern.identifier.text
+      if varName.contains(bareTypeName) {
```
Comment (allevato) [inline Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift:39]: Now that I'm getting another look at this, I wonder if we should update this to be `hasSuffix`, since the diagnostic explicitly talks about suffixes.

File: Sources/SwiftFormat/Rules/NeverForceUnwrap.swift (hunk 6 above)
Comment (shawnhyam) [reply Sources/SwiftFormat/Rules/NeverForceUnwrap.swift:19]: > Is this noise from another PR that was merged in?

This was because I ran `generate-swift-format`, and there have been some manual edits to the generated files. So I made the code edits necessary to keep the generated file the same.

File: Sources/SwiftFormat/Core/Pipelines+Generated.swift (hunk 1 above)
Comment (shawnhyam) [reply Sources/SwiftFormat/Core/Pipelines+Generated.swift:86]: @allevato I could split it out into another change, but I still want to put it in because AFAICT it's a bug in the pipeline generator (`visitPost` is never called) and I have other changes in progress that depend on that visitor method being called. Should I go ahead and split this into two separate changes?

File: Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift (hunk 7 above)
Comment (shawnhyam) [reply Sources/SwiftFormat/Rules/DontRepeatTypeInStaticProperties.swift:39]: Good question...  we wouldn't catch plurals (which we don't mention or have tests for), but there are likely false positives with the current logic too.
```
public actor Cookie {
  // cases that are caught with .contains but not with .hasSuffix
  static let chocolateChipCookies: Cookie
  static let rawCookieDough: Cookie
}
```

**Comment by allevato** [comment]
Thanks for fixing this!
